### PR TITLE
Add ability to publish/find services from loopback devices by specifying them explicitly (as well as non-loopback devices).

### PR DESCRIPTION
**Pull request by hedss**
Fix test issue where node wouldn't exist in some tests when an error occurred.

Connects to #12.

### CODE REVIEW

File: lib/discoverable.coffee [hunk 1]
```diff
@@ -153,7 +153,10 @@ isLoopbackInterface = (address) ->
 	foundIntf = _.find os.networkInterfaces(), (intfInfo) ->
 		_.find intfInfo, address: address
 
-	if not foundIntf? or  not (validFamily = _.find(foundIntf, family: family)?.internal)? then return null else return validFamily
+	if not foundIntf? or  not (validFamily = _.find(foundIntf, family: family)?.internal)?
+		return null
+
+	validFamily
```
Comment (Page-) [inline lib/discoverable.coffee]: Can you add a `return` here, I find on longer functions it is much clearer/more readable to be explicit

File: lib/discoverable.coffee [hunk 2]
```diff
@@ -153,7 +153,10 @@ isLoopbackInterface = (address) ->
 	foundIntf = _.find os.networkInterfaces(), (intfInfo) ->
 		_.find intfInfo, address: address
 
-	if not foundIntf? or  not (validFamily = _.find(foundIntf, family: family)?.internal)? then return null else return validFamily
+	if not foundIntf? or  not (validFamily = _.find(foundIntf, family: family)?.internal)?
```
Comment (Page-) [inline lib/discoverable.coffee]: There's a double space between `or  not`

Comment (hedss) [inline lib/discoverable.coffee]: Well spotted, that man!

File: lib/discoverable.coffee (hunk 1 above)
Comment (hedss) [inline lib/discoverable.coffee]: Done.

File: tests/discovery.spec.coffee [hunk 3]
```diff
@@ -257,5 +258,27 @@ describe 'Discoverable Services:', ->
...
 					discoverableServices.unpublishServices()
+
+			it 'should publish single service only to local IPv4 loopback interface', ->
+				# This is a rough check, as by default every Darwin based platform
+				# runs MDNS by default. It is assumed that every other platform (such as
+				# Linux/FreeBSD/Windows) is *not* running a Bonjour based protocol.
+				# If it is, this test will fail as it will not be able to bind to port 5354.
+				if process.platform isnt 'darwin'
+					discoverableServices.publishServices [
+						{ identifier: '_first._sub._ssh._tcp', name: 'First SSH' , port: 1234 }
```
Comment (Page-) [inline tests/discovery.spec.coffee]: There's a space before the comma in `SSH' ,`

File: lib/discoverable.coffee [hunk 4]
```diff
@@ -136,6 +137,28 @@ hasValidInterfaces = ->
...
+	if ip.isV4Format(address)
+		family = 'IPv4'
+	else if ip.isV6Format(address)
+		family = 'IPv6'
+	else
+		return null
+
+	# Correctly format any IPv6 addresses so we find them.
+	foundIntf = _.find os.networkInterfaces(), (intfInfo) ->
+		_.find intfInfo, address: address
```
Comment (Page-) [inline lib/discoverable.coffee]: This can be `_.find(intInfo, { address })`

File: tests/discovery.spec.coffee (hunk 3 above)
Comment (hedss) [inline tests/discovery.spec.coffee]: You are most definitely a space hound. Spacehound? I do wonder if we should have some linting rules for these... :)

File: lib/discoverable.coffee (hunk 4 above)
Comment (hedss) [inline lib/discoverable.coffee]: Well spotted.

**Comment by hedss** [comment]
@Page- Yes, definitely! An incredibly sensible thing to do! I've done this on a new branch to make it obvious: #26

**Comment by lekkas** [comment]
@page- amazing tip

**Comment by lekkas** [comment]
@Page- @hedss by the way, do we really need to commit the auto-generated files when we develop in languages that compile to JS (e.g. `coffeescript`, `typescript`) ? Why include the `build/` folder at all? Being able to skip suspended diffs is nice to have, but only when we trust the PR contributor.

**Comment by Page-** [comment]
Well the generated files should always be rebuilt before publishing, as to why commit them.. the only reason I can think of is to be able to install a specific commit from git, without having to build them separately?  Personally I'd be happy to only have the js files published to npm and never committed

**Comment by hedss** [comment]
My understanding was, as @Page- says, packages we pull in with git rather than `npm`. There are times when I've done this with non-published packages, as two repos essentially reference each other as WIP. It'd be annoying to lose that, but I can see the flip-side of not committing them. I'm just not sure it's worth it if we can just ignore them?

**Pull request merged**